storage/stream: fix possible unaligned write on buffer flush request

nvlsianpu opened the pull request:
On buffer flush request it is very probably that write buffer
contains amount of data which is non write-block-size aligned.
Flash memory need to be write at minimal by write-block-size chunks.
This patch addresses mechanism which ensure such behavior by adding
missing bytes.

fixes #25471

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
storage/stream: fix possible unaligned write on buffer flush

On buffer flush request it is very probably that write buffer
contains amount of data which is non write-block-size aligned.
Flash memory need to be write at minimal by write-block-size chunks.
This patch addresses mechanism which ensure such behavior by adding
missing bytes.

fixes #25471

streamer buffer size should be multiple write-block-size of
the flash device in order to avoid unaligned flash write
request.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

pabigot commented on subsys/storage/stream/stream_flash.c:
```diff
@@ -112,6 +112,8 @@ int stream_flash_buffered_write(struct stream_flash_ctx *ctx, const u8_t *data,
 	int processed = 0;
 	int rc = 0;
 	int buf_empty_bytes;
+	size_t addenum;
```
`fill_length` would be a better name than `addendum`

pabigot commented on subsys/storage/stream/stream_flash.c:
```diff
@@ -144,6 +146,25 @@ int stream_flash_buffered_write(struct stream_flash_ctx *ctx, const u8_t *data,
+			if (rc != 0) {
+				return rc;
+			}
+			memset(ctx->buf + ctx->buf_bytes, filler, addenum);
```
This is at risk for writing past the end of `ctx->buf`.  That should be avoided by adding a check in `stream_flash_init()` that `buf_len` is a multiple of the write block size.

nvlsianpu replied:
agree - was thought about such check before, but thought it is overkill

hakonfam commented on subsys/storage/stream/stream_flash.c:
```diff
@@ -112,6 +112,8 @@ int stream_flash_buffered_write(struct stream_flash_ctx *ctx, const u8_t *data,
 	int processed = 0;
 	int rc = 0;
 	int buf_empty_bytes;
+	size_t addenum;
```
I forgot to hit "send". @pabigot 's proposal is better.

nandojve commented:
I was wondering if a regression test for this would not be good.

nvlsianpu commented:
> I was wondering if a regression test for this would not be good.

Yes it is possible.

de-nordic commented on subsys/storage/stream/stream_flash.c:
```diff
@@ -189,6 +212,11 @@ int stream_flash_init(struct stream_flash_ctx *ctx, struct device *fdev,
 		return -EFAULT;
 	}
 
+	if (buf_len % flash_get_write_block_size(fdev)) {
```
Can we do that check at the beginning of `stream_flash_init`, it looks like it is parameter check.
The `buf_len` is also used at line 200, so if it is incorrect here, then it is also incorrect at that line.

de-nordic commented on subsys/storage/stream/stream_flash.c:
```diff
@@ -144,6 +146,27 @@ int stream_flash_buffered_write(struct stream_flash_ctx *ctx, const u8_t *data,
 	if (flush && ctx->buf_bytes > 0) {
+		fill_length = ctx->buf_bytes %
+				 flash_get_write_block_size(ctx->fdev);
+		if (fill_length) {
```
This is side note: We should consider adding some API call for returning the 'erased' byte value for the device. It is at least second time when I see that the value is needed and is either read from device (here) or predefined in code, e.g. in mcuboot.

de-nordic commented on subsys/storage/stream/stream_flash.c:
```diff
@@ -144,6 +146,27 @@ int stream_flash_buffered_write(struct stream_flash_ctx *ctx, const u8_t *data,
+			 * should be erased in order to get the erased
+			 * byte-value.
+			 */
+			rc = flash_read(ctx->fdev,
```
This is side note: We should consider adding some API call for returning the 'erased' byte value for the device. It is at least second time when I see that the value is needed and is either read from device (here) or predefined in code, e.g. in mcuboot.

nvlsianpu replied:
have this on my API request change list

de-nordic reviewed (approved):
Notes left. otherwise I think that the solution is fine.

nvlsianpu committed:
tests/subsys/storage/stream_flash: add unaligned flash test

Added test for check proper service of flush-write when buffer
contains unaligned amount of data.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
@nandojve Test added

pabigot commented on tests/subsys/storage/stream/stream_flash/testcase.yaml:
```diff
@@ -2,6 +2,10 @@ tests:
     platform_whitelist: native_posix native_posix_64
     tags: stream_flash
+storage.stream_flash.dword_wbs:
+    extra_args: DTC_OVERLAY=./unaligned_flush.overlay
```
Does this work?  `DTC_OVERLAY` doesn't appear to be defined.  `DTC_OVERLAY_FILE` is.

In any case using a relative path seems unusual (contrast with `OVERLAY_CONFIG=` below).

nvlsianpu replied:
fixed

hakonfam reviewed (approved):
There is a typo in the last commit "steram" -> "stream"

nandojve reviewed (approved):
After apply changes, the board disco\_l475\_iot1 started to work properly.

nvlsianpu committed:
include/storage/steram_flash: update buffer size requirements

Implementation requires that streamer buffer size must be
multiple of the flash device wbs.
This patch reflects this in API documentation

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
In the last commit added note on requirements for size of the stream buffer.
